File: ios/FilmLightMeter/Domain/DisposableCameraPreset.swift
import Foundation

public enum EvidenceLevel: String, Codable, Sendable {
    case official
    case measured
    case estimated

    public var displayName: String {
        switch self {
        case .official: "官方资料"
        case .measured: "实测资料"
        case .estimated: "经验近似"
        }
    }
}

public struct FilmProfile: Codable, Equatable, Sendable {
    public let name: String
    public let iso: Int
    public let highlightLatitudeStops: Double
    public let shadowLatitudeStops: Double
    public let evidence: EvidenceLevel

    public init(
        name: String,
        iso: Int,
        highlightLatitudeStops: Double,
        shadowLatitudeStops: Double,
        evidence: EvidenceLevel
    ) {
        precondition(iso > 0)
        precondition(highlightLatitudeStops > 0)
        precondition(shadowLatitudeStops > 0)
        self.name = name
        self.iso = iso
        self.highlightLatitudeStops = highlightLatitudeStops
        self.shadowLatitudeStops = shadowLatitudeStops
        self.evidence = evidence
    }
}

public struct FixedOptics: Codable, Equatable, Sendable {
    public let aperture: Double
    public let focalLengthMillimeters: Double
    public let minimumFocusMeters: Double?

    public init(
        aperture: Double,
        focalLengthMillimeters: Double,
        minimumFocusMeters: Double?
    ) {
        precondition(aperture > 0)
        precondition(focalLengthMillimeters > 0)
        precondition(minimumFocusMeters == nil || minimumFocusMeters! > 0)
        self.aperture = aperture
        self.focalLengthMillimeters = focalLengthMillimeters
        self.minimumFocusMeters = minimumFocusMeters
    }
}

public struct FlashProfile: Codable, Equatable, Sendable {
    public let effectiveDistanceMinMeters: Double
    public let effectiveDistanceMaxMeters: Double

    public init(
        effectiveDistanceMinMeters: Double,
        effectiveDistanceMaxMeters: Double
    ) {
        precondition(effectiveDistanceMinMeters > 0)
        precondition(effectiveDistanceMaxMeters >= effectiveDistanceMinMeters)
        self.effectiveDistanceMinMeters = effectiveDistanceMinMeters
        self.effectiveDistanceMaxMeters = effectiveDistanceMaxMeters
    }
}

public struct DisposableCameraPreset: Identifiable, Equatable, Sendable {
    public let id: String
    public let presetVersion: Int
    public let brand: String
    public let model: String
    public let regionOrBatch: String?
    public let film: FilmProfile
    public let optics: FixedOptics
    public let shutterSeconds: Double
    public let flash: FlashProfile?
    public let exposureEvidence: EvidenceLevel
    public let displayNameOverride: String?

    public init(
        id: String,
        presetVersion: Int,
        brand: String,
        model: String,
        regionOrBatch: String?,
        film: FilmProfile,
        optics: FixedOptics,
        shutterSeconds: Double,
        flash: FlashProfile?,
        exposureEvidence: EvidenceLevel,
        displayNameOverride: String? = nil
    ) {
        precondition(!id.isEmpty)
        precondition(presetVersion > 0)
        precondition(!brand.isEmpty)
        precondition(!model.isEmpty)
        precondition(shutterSeconds > 0)
        self.id = id
        self.presetVersion = presetVersion
        self.brand = brand
        self.model = model
        self.regionOrBatch = regionOrBatch
        self.film = film
        self.optics = optics
        self.shutterSeconds = shutterSeconds
        self.flash = flash
        self.exposureEvidence = exposureEvidence
        self.displayNameOverride = displayNameOverride
    }

    public var displayName: String {
        displayNameOverride ?? "\(brand) \(model)"
    }
}

public struct ManualCameraConfig: Codable, Equatable, Sendable {
    public static let presetID = "manual-camera"

    public var iso: Int
    public var shutterDenominator: Int
    public var aperture: Double
    public var focalLengthMillimeters: Double

    public init(
        iso: Int = 400,
        shutterDenominator: Int = 125,
        aperture: Double = 11,
        focalLengthMillimeters: Double = 32
    ) {
        self.iso = iso
        self.shutterDenominator = shutterDenominator
        self.aperture = aperture
        self.focalLengthMillimeters = focalLengthMillimeters
    }

    public var isValid: Bool {
        (25...6400).contains(iso)
            && (1...8000).contains(shutterDenominator)
            && aperture.isFinite && (1...64).contains(aperture)
            && focalLengthMillimeters.isFinite
            && (20...150).contains(focalLengthMillimeters)
    }

    public func makePreset() -> DisposableCameraPreset {
        precondition(isValid)
        return DisposableCameraPreset(
            id: Self.presetID,
            presetVersion: 1,
            brand: "手动",
            model: "配置",
            regionOrBatch: "自定义一次性胶片相机参数",
            film: FilmProfile(
                name: "通用彩色负片",
                iso: iso,
                highlightLatitudeStops: 3,
                shadowLatitudeStops: 2,
                evidence: .estimated
            ),
            optics: FixedOptics(
                aperture: aperture,
                focalLengthMillimeters: focalLengthMillimeters,
                minimumFocusMeters: nil
            ),
            shutterSeconds: 1 / Double(shutterDenominator),
            flash: nil,
            exposureEvidence: .estimated,
            displayNameOverride: "自定义"
        )
    }
}

public enum BuiltInDisposableCameraRepository {
    public static let presets: [DisposableCameraPreset] = [
        DisposableCameraPreset(
            id: "kodak-funsaver-800",
            presetVersion: 1,
            brand: "Kodak",
            model: "FunSaver / Fun Saver",
            regionOrBatch: "31mm · 1/100s 代表版本",
            film: FilmProfile(
                name: "Kodak ISO 800 彩色负片",
                iso: 800,
                highlightLatitudeStops: 3,
                shadowLatitudeStops: 2,
                evidence: .estimated
            ),
            optics: FixedOptics(
                aperture: 10,
                focalLengthMillimeters: 31,
                minimumFocusMeters: 1
            ),
            shutterSeconds: 1 / 100,
            flash: FlashProfile(
                effectiveDistanceMinMeters: 1.2,
                effectiveDistanceMaxMeters: 3.5
            ),
            exposureEvidence: .estimated
        ),
        DisposableCameraPreset(
            id: "kodak-power-flash-800",
            presetVersion: 1,
            brand: "Kodak",
            model: "Power Flash",
            regionOrBatch: "镜头与快门采用公开资料近似",
            film: FilmProfile(
                name: "Kodak ISO 800 彩色负片",
                iso: 800,
                highlightLatitudeStops: 3,
                shadowLatitudeStops: 2,
                evidence: .estimated
            ),
            optics: FixedOptics(
                aperture: 16,
                focalLengthMillimeters: 30,
                minimumFocusMeters: 1
            ),
            shutterSeconds: 1 / 100,
            flash: FlashProfile(
                effectiveDistanceMinMeters: 1.2,
                effectiveDistanceMaxMeters: 4.5
            ),
            exposureEvidence: .estimated
        ),
        DisposableCameraPreset(
            id: "kodak-sport-waterproof-800",
            presetVersion: 1,
            brand: "Kodak",
            model: "Sport / Waterproof",
            regionOrBatch: "镜头与快门采用同类机型保守近似",
            film: FilmProfile(
                name: "Kodak UltraMax ISO 800 彩色负片",
                iso: 800,
                highlightLatitudeStops: 3,
                shadowLatitudeStops: 2,
                evidence: .estimated
            ),
            optics: FixedOptics(
                aperture: 10,
                focalLengthMillimeters: 30,
                minimumFocusMeters: 1
            ),
            shutterSeconds: 1 / 125,
            flash: nil,
            exposureEvidence: .estimated
        ),
        DisposableCameraPreset(
            id: "fujifilm-quicksnap-flash-400",
            presetVersion: 2,
            brand: "Fujifilm",
            model: "QuickSnap Flash 400",
            regionOrBatch: "1/100s 版本",
            film: FilmProfile(
                name: "FUJICOLOR SUPERIA X-TRA 400",
                iso: 400,
                highlightLatitudeStops: 3,
                shadowLatitudeStops: 5 / 3,
                evidence: .estimated
            ),
            optics: FixedOptics(
                aperture: 10,
                focalLengthMillimeters: 32,
                minimumFocusMeters: 1
            ),
            shutterSeconds: 1 / 100,
            flash: FlashProfile(
                effectiveDistanceMinMeters: 1,
                effectiveDistanceMaxMeters: 3
            ),
            exposureEvidence: .official
        ),
        DisposableCameraPreset(
            id: "fujifilm-quicksnap-waterproof-800",
            presetVersion: 1,
            brand: "Fujifilm",
            model: "QuickSnap Waterproof / Marine",
            regionOrBatch: "ISO 800 · 10m 防水版本",
            film: FilmProfile(
                name: "Fujifilm ISO 800 彩色负片",
                iso: 800,
                highlightLatitudeStops: 3,
                shadowLatitudeStops: 2,
                evidence: .estimated
            ),
            optics: FixedOptics(
                aperture: 10,
                focalLengthMillimeters: 32,
                minimumFocusMeters: 1
            ),
            shutterSeconds: 1 / 125,
            flash: nil,
            exposureEvidence: .official
        ),
        DisposableCameraPreset(
            id: "fujifilm-c400-jelly",
            presetVersion: 1,
            brand: "Fujifilm",
            model: "C400 果冻胶卷相机",
            regionOrBatch: "预装 C400 · 可重复装卷",
            film: FilmProfile(
                name: "Fujifilm C400 ISO 400 彩色负片",
                iso: 400,
                highlightLatitudeStops: 3,
                shadowLatitudeStops: 2,
                evidence: .estimated
            ),
            optics: FixedOptics(
                aperture: 11,
                focalLengthMillimeters: 31,
                minimumFocusMeters: 1
            ),
            shutterSeconds: 1 / 125,
            flash: FlashProfile(
                effectiveDistanceMinMeters: 1,
                effectiveDistanceMaxMeters: 3
            ),
            exposureEvidence: .estimated
        ),
    ]

    public static func find(id: String, version: Int? = nil) -> DisposableCameraPreset? {
        presets.first {
            $0.id == id && (version == nil || $0.presetVersion == version)
        }
    }
}

File: android/app/src/main/java/com/lightmeter/app/filmpreview/DisposableCameraPreset.kt
package com.lightmeter.app.filmpreview

enum class EvidenceLevel(val displayName: String) {
    OFFICIAL("官方资料"),
    MEASURED("实测资料"),
    ESTIMATED("经验近似"),
}

data class FilmProfile(
    val name: String,
    val id: String = name,
    val iso: Int,
    val highlightLatitudeStops: Double,
    val shadowLatitudeStops: Double,
    val evidence: EvidenceLevel,
) {
    init {
        require(id.isNotBlank())
        require(iso > 0)
        require(highlightLatitudeStops > 0.0)
        require(shadowLatitudeStops > 0.0)
    }
}

data class FixedOptics(
    val aperture: Double,
    val focalLengthMm: Double,
    val minimumFocusMeters: Double?,
) {
    init {
        require(aperture > 0.0)
        require(focalLengthMm > 0.0)
        require(minimumFocusMeters == null || minimumFocusMeters > 0.0)
    }
}

data class FlashProfile(
    val effectiveDistanceMinMeters: Double,
    val effectiveDistanceMaxMeters: Double,
) {
    init {
        require(effectiveDistanceMinMeters > 0.0)
        require(effectiveDistanceMaxMeters >= effectiveDistanceMinMeters)
    }
}

data class DisposableCameraPreset(
    val id: String,
    val presetVersion: Int,
    val brand: String,
    val model: String,
    val regionOrBatch: String?,
    val film: FilmProfile,
    val optics: FixedOptics,
    val shutterSeconds: Double,
    val flash: FlashProfile?,
    val exposureEvidence: EvidenceLevel,
    val displayNameOverride: String? = null,
    val compatibleFilms: List<FilmProfile> = listOf(film),
) {
    init {
        require(id.isNotBlank())
        require(presetVersion > 0)
        require(brand.isNotBlank())
        require(model.isNotBlank())
        require(shutterSeconds > 0.0)
        require(compatibleFilms.isNotEmpty())
        require(compatibleFilms.any { it.id == film.id })
    }

    val displayName: String
        get() = displayNameOverride ?: "$brand $model"

    val isFilmSelectable: Boolean
        get() = compatibleFilms.size > 1

    fun withFilm(filmId: String): DisposableCameraPreset? {
        return compatibleFilms.firstOrNull { it.id == filmId }?.let { copy(film = it) }
    }
}

interface DisposableCameraRepository {
    fun presets(): List<DisposableCameraPreset>

    fun find(id: String, version: Int? = null): DisposableCameraPreset?
}

object BuiltInDisposableCameraRepository : DisposableCameraRepository {
    private const val KODAK_ULTRA_MAX_400_ID = "kodak-ultra-max-400"

    private val KODAK_EC35_FILMS = listOf(
        FilmProfile(
            id = "generic-color-100",
            name = "通用 ISO 100 彩色负片",
            iso = 100,
            highlightLatitudeStops = 3.0,
            shadowLatitudeStops = 2.0,
            evidence = EvidenceLevel.ESTIMATED,
        ),
        FilmProfile(
            id = "kodak-gold-200",
            name = "Kodak Gold 200",
            iso = 200,
            highlightLatitudeStops = 3.0,
            shadowLatitudeStops = 2.0,
            evidence = EvidenceLevel.ESTIMATED,
        ),
        FilmProfile(
            id = KODAK_ULTRA_MAX_400_ID,
            name = "Kodak Ultra Max 400",
            iso = 400,
            highlightLatitudeStops = 3.0,
            shadowLatitudeStops = 2.0,
            evidence = EvidenceLevel.ESTIMATED,
        ),
        FilmProfile(
            id = "generic-color-800",
            name = "通用 ISO 800 彩色负片",
            iso = 800,
            highlightLatitudeStops = 3.0,
            shadowLatitudeStops = 2.0,
            evidence = EvidenceLevel.ESTIMATED,
        ),
    )

    private val builtInPresets = listOf(
        DisposableCameraPreset(
            id = "kodak-power-flash-800",
            presetVersion = 3,
            brand = "Kodak",
            model = "Power Flash / FunSaver",
            regionOrBatch = "合并预设 · 31mm · f/10 · 1/100s · ISO 800",
            film = FilmProfile(
                name = "Kodak ISO 800 彩色负片",
                iso = 800,
                highlightLatitudeStops = 3.0,
                shadowLatitudeStops = 2.0,
                evidence = EvidenceLevel.ESTIMATED,
            ),
            optics = FixedOptics(
                aperture = 10.0,
                focalLengthMm = 31.0,
                minimumFocusMeters = 1.0,
            ),
            shutterSeconds = 1.0 / 100.0,
            flash = FlashProfile(
                effectiveDistanceMinMeters = 1.2,
                effectiveDistanceMaxMeters = 3.5,
            ),
            exposureEvidence = EvidenceLevel.ESTIMATED,
        ),
        DisposableCameraPreset(
            id = "fujifilm-quicksnap-flash-400",
            presetVersion = 3,
            brand = "Fujifilm",
            model = "QuickSnap Flash 400",
            regionOrBatch = "32mm · f/10 · 1/140s · ISO 400",
            film = FilmProfile(
                name = "FUJICOLOR SUPERIA X-TRA 400",
                iso = 400,
                highlightLatitudeStops = 3.0,
                shadowLatitudeStops = 5.0 / 3.0,
                evidence = EvidenceLevel.ESTIMATED,
            ),
            optics = FixedOptics(
                aperture = 10.0,
                focalLengthMm = 32.0,
                minimumFocusMeters = 1.0,
            ),
            shutterSeconds = 1.0 / 140.0,
            flash = FlashProfile(
                effectiveDistanceMinMeters = 1.0,
                effectiveDistanceMaxMeters = 3.0,
            ),
            exposureEvidence = EvidenceLevel.OFFICIAL,
        ),
        DisposableCameraPreset(
            id = "fujifilm-c400-jelly",
            presetVersion = 2,
            brand = "Fujifilm",
            model = "C400 果冻胶卷相机",
            regionOrBatch = "32mm · f/11 · 1/125s · ISO 400 · 单次 36 张",
            film = FilmProfile(
                name = "Fujifilm C400 ISO 400 彩色负片",
                iso = 400,
                highlightLatitudeStops = 3.0,
                shadowLatitudeStops = 2.0,
                evidence = EvidenceLevel.ESTIMATED,
            ),
            optics = FixedOptics(
                aperture = 11.0,
                focalLengthMm = 32.0,
                minimumFocusMeters = 1.0,
            ),
            shutterSeconds = 1.0 / 125.0,
            flash = FlashProfile(
                effectiveDistanceMinMeters = 1.0,
                effectiveDistanceMaxMeters = 3.0,
            ),
            exposureEvidence = EvidenceLevel.ESTIMATED,
        ),
        DisposableCameraPreset(
            id = "kodak-ec35-reusable",
            presetVersion = 1,
            brand = "Kodak",
            model = "EC35",
            regionOrBatch = "25mm · f/10 · 1/100s · 可换 135 胶卷",
            film = KODAK_EC35_FILMS.first { it.id == KODAK_ULTRA_MAX_400_ID },
            optics = FixedOptics(
                aperture = 10.0,
                focalLengthMm = 25.0,
                minimumFocusMeters = 1.0,
            ),
            shutterSeconds = 1.0 / 100.0,
            flash = null,
            exposureEvidence = EvidenceLevel.ESTIMATED,
            compatibleFilms = KODAK_EC35_FILMS,
        ),
    )

    override fun presets(): List<DisposableCameraPreset> = builtInPresets

    override fun find(id: String, version: Int?): DisposableCameraPreset? {
        return builtInPresets.firstOrNull {
            it.id == id && (version == null || it.presetVersion == version)
        }
    }

}
